nixos/tests/ec2-image.nix

```diff
@@ -0,0 +1,266 @@
+
+          # In a NixOS test the serial console is occupied by the "backdoor"
+          # (see testing/test-instrumentation.nix) and is incompatible with
+          # the configuration in virtualisation/amazon-image.nix.
+          systemd.services."serial-getty@ttyS0".enable = mkForce false;
+
+          # Make /dev/console point to serial console for proper capture
+          # test-instrumentation.nix adds console=tty0, so we need to add console=ttyS0 AFTER it
+          # The last console= parameter takes precedence for /dev/console
+          boot.kernelParams = mkAfter [ "console=ttyS0" ];
+
+          # Configure VLAN networking to match test framework setup
+          networking.interfaces.eth0 = {
+            ipv4.addresses = [
+              {
+                address = config.nodes.machine.networking.primaryIPAddress;
+                prefixLength = 24;
+              }
+            ];
+          };
+
+          nixpkgs.pkgs = pkgs;
+        }
+      ];
+    }).config;
+
+  image = "${imageCfg.system.build.amazonImage}/${imageCfg.image.fileName}";
+
+in
+{
+  name = "ec2-image";
+  meta = {
+    maintainers = with lib.maintainers; [
+      roberth
+      arianvp
+    ];
+    timeout = 600;
+  };
+  skipTypeCheck = true; # Disable type checking due to MAC address parsing issues
+  skipLint = true; # Disable linting due to MAC address parsing issues
+
+  nodes = {
+    machine = { ... }: { }; # Dummy node for network config - won't be launched
+    client =
+      { ... }:
+      {
+        # Configure SSH client for non-interactive, strict authentication
+        programs.ssh.extraConfig = ''
+          Host *
+            PasswordAuthentication no
+            ChallengeResponseAuthentication no
+            HostbasedAuthentication no
+            BatchMode yes
+            PubkeyAuthentication yes
+            StrictHostKeyChecking yes
+        '';
+      };
+  };
+
+  testScript = ''
+    import os
+    import re
+    import subprocess
+    import tempfile
+
+    # Instance Metadata Service
+    # FIXME: Support IMDSv2
+    #        - Consider https://github.com/aws/amazon-ec2-metadata-mock, or
+    #        - Consider https://github.com/purpleclay/imds-mock
+    #        - Use IMDSv2 to provide the client's pubkey to the instance
+    #        - [Test matrix] also test providing the host key through IMDS
+    #          - i.e. a test module argument to select between writing or reading the host key
+    def create_ec2_metadata_dir(temp_dir, client_pubkey):
+        """Create fake EC2 metadata directory structure with mock data"""
+        metadata_dir = os.path.join(temp_dir.name, "ec2-metadata")
+
+        # Create directory structure
+        os.makedirs(os.path.join(metadata_dir, "1.0", "meta-data", "public-keys", "0"), exist_ok=True)
+        os.makedirs(os.path.join(metadata_dir, "latest", "api"), exist_ok=True)
+
+        # IMDS v1 endpoints (what fetch-ec2-metadata.sh actually fetches)
```

Comment:
You're confusing concepts here.   IMDSv1 has to do with the authentication mechanism; not with the `1.0` uri.    IMDSv2 also has the 1.0 endpoints.  it just doesn't allow you to access the URI without a token.

We dont' support IMDSv1 in the NixOS images at all anymore

Reply:
What you are implementing here is indeed IMDSv2. But with the  `1.0` version of the metadata

Reply:
[Updated](https://github.com/NixOS/nixpkgs/compare/b889f0ba4ca825fb7eef29999f194fc1d7403277..e88783f9685a2e31e1da170d5e0b96220ec421ed) and added TODO for the latest version